In [1]:
import pandas as pd


In [2]:
dataset=pd.read_csv("Placement.csv")
dataset

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0        1      M  67.00   Others  91.00   Others  Commerce     58.00   
1        2      M  79.33  Central  78.33   Others   Science     77.48   
2        3      M  65.00  Central  68.00  Central      Arts     64.00   
3        4      M  56.00  Central  52.00  Central   Science     52.00   
4        5      M  85.80  Central  73.60  Central  Commerce     73.30   
..     ...    ...    ...      ...    ...      ...       ...       ...   
210    211      M  80.60   Others  82.00   Others  Commerce     77.60   
211    212      M  58.00   Others  60.00   Others   Science     72.00   
212    213      M  67.00   Others  67.00   Others  Commerce     73.00   
213    214      F  74.00   Others  66.00   Others  Commerce     58.00   
214    215      M  62.00  Central  58.00   Others   Science     53.00   

      degree_t workex  etest_p specialisation  mba_p      status    salary  
0     Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1     Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2    Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3     Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4    Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0  
..         ...    ...      ...            ...    ...         ...       ...  
210  Comm&Mgmt     No     91.0        Mkt&Fin  74.49      Placed  400000.0  
211   Sci&Tech     No     74.0        Mkt&Fin  53.62      Placed  275000.0  
212  Comm&Mgmt    Yes     59.0        Mkt&Fin  69.72      Placed  295000.0  
213  Comm&Mgmt     No     70.0         Mkt&HR  60.23      Placed  204000.0  
214  Comm&Mgmt     No     89.0         Mkt&HR  60.22  Not Placed       NaN  

[215 rows x 15 columns]

In [3]:
from Univariate import UnivariateBuilder
quan,qual = UnivariateBuilder.quanQualFinder(dataset)
print("quan:",quan)
print("qual:",qual)

quan: ['sl_no', 'ssc_p', 'hsc_p', 'degree_p', 'etest_p', 'mba_p', 'salary']
qual: ['gender', 'ssc_b', 'hsc_b', 'hsc_s', 'degree_t', 'workex', 'specialisation', 'status']


In [4]:
descriptive = pd.DataFrame()
descriptive

Empty DataFrame
Columns: []
Index: []

In [5]:
import numpy as np
# describe = np.percentile(dataset,10)
dataset.describe()

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

In [6]:
descriptive = pd.DataFrame(index = ["Mean","Median","Mode","Q1:25%",
                                    "Q2:50%","Q3:75%","98%","Q4:100%","IQR","1.5Rule","Lesser","Grather","Min","Max"                               
                                   ],columns=quan)
descriptive

sl_no ssc_p hsc_p degree_p etest_p mba_p salary
Mean      NaN   NaN   NaN      NaN     NaN   NaN    NaN
Median    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Mode      NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q1:25%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q2:50%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q3:75%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
98%       NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q4:100%   NaN   NaN   NaN      NaN     NaN   NaN    NaN
IQR       NaN   NaN   NaN      NaN     NaN   NaN    NaN
1.5Rule   NaN   NaN   NaN      NaN     NaN   NaN    NaN
Lesser    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Grather   NaN   NaN   NaN      NaN     NaN   NaN    NaN
Min       NaN   NaN   NaN      NaN     NaN   NaN    NaN
Max       NaN   NaN   NaN      NaN     NaN   NaN    NaN

In [7]:
import numpy as np
import warnings
warnings.filterwarnings("ignore")

for colName in quan:
    descriptive[colName]['Mean'] = dataset[colName].mean()
    descriptive[colName]['Median'] = dataset[colName].median()
    descriptive[colName]['Mode'] = dataset[colName].mode()[0]
    descriptive[colName]['Q1:25%'] = dataset.describe()[colName]["25%"]
    descriptive[colName]['Q2:50%'] = dataset.describe()[colName]["50%"]
    descriptive[colName]['Q3:75%'] = dataset.describe()[colName]["75%"]
    # descriptive[colName]['98%'] = np.percentile(dataset[colName],98)
    descriptive[colName]['Q4:100%'] = dataset.describe()[colName]["max"]
    descriptive[colName]['IQR'] = descriptive[colName]['Q3:75%'] - descriptive[colName]['Q1:25%']
    descriptive[colName]['1.5Rule'] = 1.5 * descriptive[colName]['IQR']
    descriptive[colName]['Lesser'] = descriptive[colName]['Q1:25%'] - descriptive[colName]['1.5Rule'] #Q1-1.5*IQR = Q1-descriptive[colName]['1.5Rule']
    descriptive[colName]['Grather'] = descriptive[colName]['Q3:75%'] + descriptive[colName]['1.5Rule'] #Q3+1.5*IQR = Q1-descriptive[colName]['1.5Rule']
    descriptive[colName]['Min'] = dataset[colName].min() #Min values have to calc from dataset
    descriptive[colName]['Max'] = dataset[colName].max() #Max values have to calc from dataset
    

In [8]:
descriptive

sl_no      ssc_p      hsc_p   degree_p    etest_p      mba_p  \
Mean     108.0  67.303395  66.333163  66.370186  72.100558  62.278186   
Median   108.0       67.0       65.0       66.0       71.0       62.0   
Mode         1       62.0       63.0       65.0       60.0       56.7   
Q1:25%    54.5       60.6       60.9       61.0       60.0     57.945   
Q2:50%   108.0       67.0       65.0       66.0       71.0       62.0   
Q3:75%   161.5       75.7       73.0       72.0       83.5     66.255   
98%        NaN        NaN        NaN        NaN        NaN        NaN   
Q4:100%  215.0       89.4       97.7       91.0       98.0      77.89   
IQR      107.0       15.1       12.1       11.0       23.5       8.31   
1.5Rule  160.5      22.65      18.15       16.5      35.25     12.465   
Lesser  -106.0      37.95      42.75       44.5      24.75      45.48   
Grather  322.0      98.35      91.15       88.5     118.75      78.72   
Min          1      40.89       37.0       50.0       50.0      51.21   
Max        215       89.4       97.7       91.0       98.0      77.89   

                salary  
Mean     288655.405405  
Median        265000.0  
Mode          300000.0  
Q1:25%        240000.0  
Q2:50%        265000.0  
Q3:75%        300000.0  
98%                NaN  
Q4:100%       940000.0  
IQR            60000.0  
1.5Rule        90000.0  
Lesser        150000.0  
Grather       390000.0  
Min           200000.0  
Max           940000.0

### Find Outlier columns

In [9]:
lessorCol=[]
greaterCol=[]

for colName in quan:
    if(descriptive[colName]["Lesser"] >  descriptive[colName]["Min"]):
        lessorCol.append(colName)

    if(descriptive[colName]["Grather"] <  descriptive[colName]["Max"]):
     greaterCol.append(colName)

In [10]:
print('lessorCol',lessorCol)
print('greaterCol',greaterCol)

lessorCol ['hsc_p']
greaterCol ['hsc_p', 'degree_p', 'salary']


### Replacing Outliers

In [11]:
for colName in lessorCol:
    dataset[colName][dataset[colName] < descriptive[colName]["Lesser"]] =  descriptive[colName]["Lesser"]
 
for colName in greaterCol:
    dataset[colName][dataset[colName] > descriptive[colName]["Grather"]] =  descriptive[colName]["Grather"]   
    
    
    

In [12]:
# dataset

In [13]:
descriptive = pd.DataFrame(index = ["Mean","Median","Mode","Q1:25%",
                                    "Q2:50%","Q3:75%","98%","Q4:100%","IQR","1.5Rule","Lesser","Grather","Min","Max"                               
                                   ],columns=quan)
descriptive

sl_no ssc_p hsc_p degree_p etest_p mba_p salary
Mean      NaN   NaN   NaN      NaN     NaN   NaN    NaN
Median    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Mode      NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q1:25%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q2:50%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q3:75%    NaN   NaN   NaN      NaN     NaN   NaN    NaN
98%       NaN   NaN   NaN      NaN     NaN   NaN    NaN
Q4:100%   NaN   NaN   NaN      NaN     NaN   NaN    NaN
IQR       NaN   NaN   NaN      NaN     NaN   NaN    NaN
1.5Rule   NaN   NaN   NaN      NaN     NaN   NaN    NaN
Lesser    NaN   NaN   NaN      NaN     NaN   NaN    NaN
Grather   NaN   NaN   NaN      NaN     NaN   NaN    NaN
Min       NaN   NaN   NaN      NaN     NaN   NaN    NaN
Max       NaN   NaN   NaN      NaN     NaN   NaN    NaN

In [14]:
import numpy as np
import warnings
warnings.filterwarnings("ignore")

for colName in quan:
    descriptive[colName]['Mean'] = dataset[colName].mean()
    descriptive[colName]['Median'] = dataset[colName].median()
    descriptive[colName]['Mode'] = dataset[colName].mode()[0]
    descriptive[colName]['Q1:25%'] = dataset.describe()[colName]["25%"]
    descriptive[colName]['Q2:50%'] = dataset.describe()[colName]["50%"]
    descriptive[colName]['Q3:75%'] = dataset.describe()[colName]["75%"]
    # descriptive[colName]['98%'] = np.percentile(dataset[colName],98)
    descriptive[colName]['Q4:100%'] = dataset.describe()[colName]["max"]
    descriptive[colName]['IQR'] = descriptive[colName]['Q3:75%'] - descriptive[colName]['Q1:25%']
    descriptive[colName]['1.5Rule'] = 1.5 * descriptive[colName]['IQR']
    descriptive[colName]['Lesser'] = descriptive[colName]['Q1:25%'] - descriptive[colName]['1.5Rule'] #Q1-1.5*IQR = Q1-descriptive[colName]['1.5Rule']
    descriptive[colName]['Grather'] = descriptive[colName]['Q3:75%'] + descriptive[colName]['1.5Rule'] #Q3+1.5*IQR = Q1-descriptive[colName]['1.5Rule']
    descriptive[colName]['Min'] = dataset[colName].min() #Min values have to calc from dataset
    descriptive[colName]['Max'] = dataset[colName].max() #Max values have to calc from dataset

### After Replacing outliers

In [15]:
lessorCol=[]
greaterCol=[]

for colName in quan:
    if(descriptive[colName]["Lesser"] >  descriptive[colName]["Min"]):
        lessorCol.append(colName)

    if(descriptive[colName]["Grather"] <  descriptive[colName]["Max"]):
     greaterCol.append(colName)

In [16]:
print('lessorCol',lessorCol)
print('greaterCol',greaterCol)

lessorCol []
greaterCol []


### Create Frequency Table

In [25]:
def createFreqTable(dataset,colName):
    freqTable = pd.DataFrame(columns=['Unique_values','Frequency','Relative_Frequency',"CumSum"])
    freqTable['Unique_values']=dataset[colName].value_counts().index
    freqTable['Frequency']=dataset[colName].value_counts().values
    freqTable['Relative_Frequency']=( freqTable['Frequency'] / len(dataset[colName].value_counts()))
    freqTable['CumSum']=freqTable['Relative_Frequency'].cumsum()
    return freqTable

In [26]:
createFreqTable(dataset,colName)

Unique_values  Frequency  Relative_Frequency    CumSum
0        300000.0         22            0.611111  0.611111
1        250000.0         18            0.500000  1.111111
2        390000.0         15            0.416667  1.527778
3        240000.0         15            0.416667  1.944444
4        260000.0          7            0.194444  2.138889
5        200000.0          6            0.166667  2.305556
6        265000.0          6            0.166667  2.472222
7        360000.0          6            0.166667  2.638889
8        275000.0          5            0.138889  2.777778
9        220000.0          5            0.138889  2.916667
10       270000.0          4            0.111111  3.027778
11       210000.0          4            0.111111  3.138889
12       350000.0          3            0.083333  3.222222
13       216000.0          3            0.083333  3.305556
14       204000.0          2            0.055556  3.361111
15       236000.0          2            0.055556  3.416667
16       218000.0          2            0.055556  3.472222
17       230000.0          2            0.055556  3.527778
18       252000.0          2            0.055556  3.583333
19       280000.0          2            0.055556  3.638889
20       276000.0          2            0.055556  3.694444
21       278000.0          1            0.027778  3.722222
22       233000.0          1            0.027778  3.750000
23       264000.0          1            0.027778  3.777778
24       290000.0          1            0.027778  3.805556
25       285000.0          1            0.027778  3.833333
26       255000.0          1            0.027778  3.861111
27       340000.0          1            0.027778  3.888889
28       380000.0          1            0.027778  3.916667
29       225000.0          1            0.027778  3.944444
30       320000.0          1            0.027778  3.972222
31       336000.0          1            0.027778  4.000000
32       231000.0          1            0.027778  4.027778
33       268000.0          1            0.027778  4.055556
34       287000.0          1            0.027778  4.083333
35       295000.0          1            0.027778  4.111111

In [44]:
def Univariate(dataset,quan):
    descriptive = pd.DataFrame(index = ["Mean","Median","Mode","Q1:25%",
                                    "Q2:50%","Q3:75%","99%","Q4:100%","IQR","1.5Rule","Lesser","Grather","Min","Max",'skew','kurtosis'                               
                                   ],columns=quan)
    for colName in quan:
        descriptive[colName]['Mean'] = dataset[colName].mean()
        descriptive[colName]['Median'] = dataset[colName].median()
        descriptive[colName]['Mode'] = dataset[colName].mode()[0]
        descriptive[colName]['Q1:25%'] = dataset.describe()[colName]["25%"]
        descriptive[colName]['Q2:50%'] = dataset.describe()[colName]["50%"]
        descriptive[colName]['Q3:75%'] = dataset.describe()[colName]["75%"]
        descriptive[colName]['99%'] = np.percentile(dataset[colName],99)
        descriptive[colName]['Q4:100%'] = dataset.describe()[colName]["max"]
        descriptive[colName]['IQR'] = descriptive[colName]['Q3:75%'] - descriptive[colName]['Q1:25%']
        descriptive[colName]['1.5Rule'] = 1.5 * descriptive[colName]['IQR']
        descriptive[colName]['Lesser'] = descriptive[colName]['Q1:25%'] - descriptive[colName]['1.5Rule'] #Q1-1.5*IQR = Q1-descriptive[colName]['1.5Rule']
        descriptive[colName]['Grather'] = descriptive[colName]['Q3:75%'] + descriptive[colName]['1.5Rule'] #Q3+1.5*IQR = Q1-descriptive[colName]['1.5Rule']
        descriptive[colName]['Min'] = dataset[colName].min() #Min values have to calc from dataset
        descriptive[colName]['Max'] = dataset[colName].max() #Max values have to calc from dataset
        descriptive[colName]['skew'] = dataset[colName].skew()
        descriptive[colName]['kurtosis'] = dataset[colName].kurtosis()
    return descriptive

Univariate(dataset,quan)

## Create Reusable function FindOutlierCol

In [34]:
def FindOutlierCol(quan):
    lessorCol=[]
    greaterCol=[]

    for colName in quan:
        if(descriptive[colName]["Lesser"] >  descriptive[colName]["Min"]):
            lessorCol.append(colName)
    
        if(descriptive[colName]["Grather"] <  descriptive[colName]["Max"]):
         greaterCol.append(colName)

    return lessorCol,greaterCol
    

In [35]:
 FindOutlierCol(quan)

([], [])

## Create Reusable function ReplacingOutliers

In [41]:
def ReplacingOutliers(dataset,lessorCol,greaterCol):
    for colName in lessorCol:
        dataset[colName][dataset[colName] < descriptive[colName]["Lesser"]] =  descriptive[colName]["Lesser"]
     
    for colName in greaterCol:
        dataset[colName][dataset[colName] > descriptive[colName]["Grather"]] =  descriptive[colName]["Grather"] 
    return dataset

In [42]:
ReplacingOutliers(dataset,lessorCol,greaterCol)

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0        1      M  67.00   Others  91.00   Others  Commerce     58.00   
1        2      M  79.33  Central  78.33   Others   Science     77.48   
2        3      M  65.00  Central  68.00  Central      Arts     64.00   
3        4      M  56.00  Central  52.00  Central   Science     52.00   
4        5      M  85.80  Central  73.60  Central  Commerce     73.30   
..     ...    ...    ...      ...    ...      ...       ...       ...   
210    211      M  80.60   Others  82.00   Others  Commerce     77.60   
211    212      M  58.00   Others  60.00   Others   Science     72.00   
212    213      M  67.00   Others  67.00   Others  Commerce     73.00   
213    214      F  74.00   Others  66.00   Others  Commerce     58.00   
214    215      M  62.00  Central  58.00   Others   Science     53.00   

      degree_t workex  etest_p specialisation  mba_p      status    salary  
0     Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1     Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2    Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3     Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4    Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  390000.0  
..         ...    ...      ...            ...    ...         ...       ...  
210  Comm&Mgmt     No     91.0        Mkt&Fin  74.49      Placed  390000.0  
211   Sci&Tech     No     74.0        Mkt&Fin  53.62      Placed  275000.0  
212  Comm&Mgmt    Yes     59.0        Mkt&Fin  69.72      Placed  295000.0  
213  Comm&Mgmt     No     70.0         Mkt&HR  60.23      Placed  204000.0  
214  Comm&Mgmt     No     89.0         Mkt&HR  60.22  Not Placed       NaN  

[215 rows x 15 columns]

In [47]:
Univariate(dataset,quan)

sl_no      ssc_p      hsc_p   degree_p    etest_p      mba_p  \
Mean       108.0  67.303395  66.334744  66.358558  72.100558  62.278186   
Median     108.0       67.0       65.0       66.0       71.0       62.0   
Mode           1       62.0       63.0       65.0       60.0       56.7   
Q1:25%      54.5       60.6       60.9       61.0       60.0     57.945   
Q2:50%     108.0       67.0       65.0       66.0       71.0       62.0   
Q3:75%     161.5       75.7       73.0       72.0       83.5     66.255   
99%       212.86       87.0     91.129      83.86       97.0    76.1142   
Q4:100%    215.0       89.4      91.15       88.5       98.0      77.89   
IQR        107.0       15.1       12.1       11.0       23.5       8.31   
1.5Rule    160.5      22.65      18.15       16.5      35.25     12.465   
Lesser    -106.0      37.95      42.75       44.5      24.75      45.48   
Grather    322.0      98.35      91.15       88.5     118.75      78.72   
Min            1      40.89      42.75       50.0       50.0      51.21   
Max          215       89.4      91.15       88.5       98.0      77.89   
skew         0.0  -0.132649   0.162611   0.204164   0.282308   0.313576   
kurtosis    -1.2   -0.60751   0.086901   -0.09749   -1.08858  -0.470723   

                 salary  
Mean      277648.648649  
Median         265000.0  
Mode           300000.0  
Q1:25%         240000.0  
Q2:50%         265000.0  
Q3:75%         300000.0  
99%                 NaN  
Q4:100%        390000.0  
IQR             60000.0  
1.5Rule         90000.0  
Lesser         150000.0  
Grather        390000.0  
Min            200000.0  
Max            390000.0  
skew             0.8067  
kurtosis      -0.239837

### Understood from POC 

|  Skewness Value | Short Name                     | Meaning (Simple)                                                           |
| --------------- | ------------------------------ | -------------------------------------------------------------------------- |
|s = 0            | **Normal**                     | Data is balanced on both sides                                             |
|s > 0            | **Positive Skew** (Right Skew) | Tail on the **right** side is longer (more small values, few large values) |
|s < 0            | **Negative Skew** (Left Skew)  | Tail on the **left** side is longer (more large values, few small values)  |


Skewness sl_no: 0.0 <b>Normal</b><br>
Skewness ssc_p: -0.132649 <b>Positive</b><br>
Skewness hsc_p: 0.162611 <b>Negative</b><br>
Skewness degree_p: 0.204164 <b>Negative</b><br>
Skewness etest_p: 0.282308	<b>Negative</b><br>
Skewness mba_p: 0.313576 <b>Negative</b><br>
Skewness salary: 0.8067 <b>Negative</b><br>

|  Kurtosis Value     | Short Name      | Meaning (Simple)                           |
| ---------------     | --------------- | ------------------------------------------ |
|k = 3 or 0           | **Mesokurtic**  | Normal height and tails                    |
|k > 3 or 0           | **Leptokurtic** | High peak and heavy tails (more outliers)  |
|k < 3 or 0           | **Platykurtic** | Flat peak and light tails (fewer outliers) |


Kurtosis sl_no: -1.2 <b>light</b><br>
Kurtosis ssc_p: -0.60751 <b>light</b><br>
Kurtosis hsc_p: 0.086901 <b>heavy</b><br>
Kurtosis degree_p: -0.09749 <b>light</b><br>
Kurtosis etest_p: -1.08858	<b>light</b><br>
Kurtosis mba_p: -0.470723 <b>light</b><br>
Kurtosis salary: -0.239837 <b>light</b><br>